### ISSUE TITLE
Bug: no-unnecessary-type-parameters false positive

### ISSUE THREAD

**Issue by QuixThe2nd** [typescript-eslint/typescript-eslint#11295]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.8.2&fileType=.tsx&code=MYewdgzgLgBAtgTwGIFczCgS3DAvDAHgBkBDAIwFMAbGCgDygrABMIZoAnTMAcwD4AFFXLUAXDFKUqASjx8YAbwBQMGKEixhUgEzjJ1PDC3UA3CrXgIIKhQB0VEDyEiq26WYC%2BQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJCXQO0SODABfELqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const myFunction = <Label extends string>(label: Label) => {
  const label2: Label = label;
  console.log(label2);
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-parameters": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

When a type parameter is used only once in the function signature, but also referenced elsewhere in the function, the error should not be thrown.

### Actual Result

tseslint doesn't seem to detect the second reference to the type outside the signature.

**Comment by bradzacher** [comment]
This is working as expected - you have created a generic in the function signature for an implementation detail in the function body.

You could update the code to enclose the detail so it does not leak into the signature:

```ts
const myFunction = (label: string) => {
  type Label = typeof label;
  const label2: Label = label;
  console.log(label2);
}
```
